Ideas on Quarkus command-mode, UI-based apps and releasing via Github Actions (#546)
quarkusio/quarkusio.github.io: _posts/2020-05-01-native-command-mode-tools.adoc

---
layout: post
title: 'Ideas on Quarkus command-mode, UI-based apps and releasing via Github Actions'
date: 2020-05-10
tags: development-tips
synopsis: Ideas on how to use Quarkus command-mode to develop native applications with web frontends; built with Github Actions.
author: dplatz
---
:imagesdir: /assets/images/posts/quarkus-native-app

> Quarkus is advertised as a cloud-native runtime for Microservices but the link:https://quarkus.io/blog/quarkus-1-4-final-released/[introduction of command-mode] might have opened up new and intresting spaces where it could find adoption.

Reviewer: ```suggestion
Quarkus is advertised as a cloud-native runtime for Microservices but the link:https://quarkus.io/blog/quarkus-1-4-final-released/[introduction of command-mode] might have opened up new and interesting spaces where it could find adoption.
```

In this blog post I will present a few ideas about the possible usage of command-mode in Quarkus that you might find useful to consider when developing your next tool.
If you have not heard about command-mode, I suggest to start reading link:https://quarkus.io/blog/introducing-command-mode/[this introduction] first and then come back.

Ready? Ok, let's continue.

== Cross-platform builds via Github Actions

Command-line tools like `kubectl` are often written in languages like C, Go or Rust and compiled to native executables for various target platforms.
This is great because you don't need to rely on Java installed; or beeing installed in a wrong version.

> With GraalVM native-images, this argument against Java does not hold anymore. Quarkus supports the build of native executables via GraalVM. Using it in combination with command-mode sounds a very productive choice to write command-line tools.

Reviewer: might want to mention that Quarkus supports mainly building natively for Linux containers, but also has experimental support for OS X and Windows or something like that

Author: Rephrased it to:
With GraalVM native-images, this argument against Java does not hold anymore. Quarkus supports the build of native executables via GraalVM; mainly for building natively for Linux containers but there is experimental support for OS X and Windows as well.
Using the native build in combination with command-mode sounds like a very interesting and productive choice to write command-line tools.

    - name: Build native executable
      run: ./mvnw package -Dnative <4>
    - name: Upload native executable
      id: upload-native-executable
      uses: actions/upload-release-asset@v1
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        upload_url: ${{ github.event.release.upload_url }}
        asset_path: ./target/quarkus-ls-${{ github.event.release.tag_name }}-runner <5>
        asset_name: quarkus-ls-${{ github.event.release.tag_name }}-linux
        asset_content_type: application/octet-stream
----
<1> This build-job is triggered when a new release is triggered. If your are the owner, you will have the button "Draft a new release" under link:https://github.com/38leinaD/quarkus-native-app-starter/releases[releases]
<2> Uses a ubuntu runner. See link:https://github.com/actions/virtual-environments[GitHub Actions Virtual Environments] for a list of all supported environments/runners.
<3> Set version in `pom.xml` based on specified release version
<4> Build the native executable
<5> Upload/attach the artifact to the release

Building for Mac just means to provide another job with `runs-on: macos-latest`.

> Windows is essentially the same but requires a little more attention and knowledge about the tooling. Getting it working required some investigation on how the native build depends on Visual Studio. Also,the Windows runners on Github Actions have a too small page-file by default. The memory consumption is rather high when running the GraalVMs `native-image` command.

Reviewer: ```suggestion
Windows is essentially the same but requires a little more attention and knowledge about the tooling. Getting it working required some investigation on how the native build depends on Visual Studio. Also,the Windows runners on Github Actions have a too small page-file by default. That needs increasing as the memory consumption is rather high when running the GraalVMs `native-image` command.
```

An full example of a working native-image build under windows can be found link:https://github.com/38leinaD/quarkus-native-app-starter/blob/master/.github/workflows/release-build-native-win64.yml[here].

After triggering a release from the Github UI, the jobs for all target environments are triggered. The artifacts for all platforms are will be available within minutes and can be consumed by your users. Building, hosting: 100% free and no external service or registration needed. 

image:github-actions.jpg[]

== Command-mode for UI-based applications

Another exciting thing about command-mode is that it not only allows to write nice command-line tools (using e.g. link:https://picocli.info/[Picocli], link:http://aeshell.github.io/[Aesh] or link:https://commons.apache.org/proper/commons-cli/[Apache Commons CLI]) but in general allows to have more control over the lifecycle of an application.

> With the possibiliy to embeed your HTML/Javascript frontend, we can even build link:https://www.electronjs.org/[Electron-style] desktop application with rich user interfaces for our tools. As even JSF most likely will be link:https://github.com/adminfaces/quarkus-admin-starter[coming to Quarkus], there is a great variety of choices how to build a frontend. 

Reviewer: ```suggestion
With the possibility to embed your HTML/Javascript frontend, we can even build link:https://www.electronjs.org/[Electron-style] desktop application with rich user interfaces for our tools. As even JSF most likely will be link:https://github.com/adminfaces/quarkus-admin-starter[coming to Quarkus], there is a great variety of choices how to build a frontend. 
```

> A very simple scaffold does not require much. If you have `quarkus-resteasy` as a dependency, the embeeded server will be started on port `8080` just like in any regular Quarkus backend application.

Reviewer: ```suggestion
A very simple scaffold does not require much. If you have `quarkus-resteasy` as a dependency, the embedded server will be started on port `8080` just like in any regular Quarkus backend application.
```

As your tools might be running on a developer system alongside other app-servers or Quarkus instances, it is a good idea to choose a different port to not run into conflicts.

Configuring `quarkus.http.port=0` in your `application.properties` will make Quarkus choose a random but free port.
How to know from within your application which port it is? When injecting the config-property via `@ConfigProperty(name = "quarkus.http.port")`, you will get the assigned port.

> Assuming you have your front-end under `src/main/resources/META-INF/resources` (in my example it is plain HTML with Javascript) and using a very old Java API, you can spin up the default browser to your app like this: 

Reviewer: which "old java api" is that referring to ? :)

Author: Desktop.getDesktop().browse :-)
i have added it to the text to be clearer

[source, java]
----
@QuarkusMain
public class HelloMain implements QuarkusApplication {

    @ConfigProperty(name = "quarkus.http.port")
    Integer assignedPort; <1>
    
    public static void main(String[] args) {
        Quarkus.run(HelloMain.class, args);
    }

    @Override
    public int run(String... args) throws Exception {
        URI webappUri = new URI("http://localhost:" + assignedPort + "/index.html");
        

>         Desktop.getDesktop().browse(webappUri);

Reviewer: might want to add the key imports  on top of this..

Author: Done
